sandbox/gpio/common.go

```diff
@@ -0,0 +1,276 @@
+	if err := os.MkdirAll(filepath.Join(dirs.GlobalRootDir, ephemeralUdevRulesDir), 0755); err != nil {
+		return err
+	}
+
+	tag := fmt.Sprintf("snap_%s_interface_gpio_chardev_%s", instanceName, slotName)
+	rule := fmt.Sprintf("SUBSYSTEM==\"gpio\", KERNEL==\"%s\", TAG+=\"%s\"\n", chip.Name, tag)
+
+	path := aggregatedChipUdevRulePath(instanceName, slotName)
+	if err := os.WriteFile(path, []byte(rule), 0644); err != nil {
+		return err
+	}
+
+	// make sure the rule we just dropped is loaded as sometimes it doesn't get
+	// picked up right away
+	output, err := exec.CommandContext(ctx, "udevadm", "control", "--reload-rules").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cannot reload udev rules: %s\nudev output:\n%s", err, string(output))
+	}
+	// trigger the tagging rule
+	output, err = exec.CommandContext(ctx, "udevadm", "trigger", "--name-match", chip.Name).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cannot trigger udev rules: %s\nudev output:\n%s", err, string(output))
+	}
+
+	return nil
+}
+
+var unixStat = unix.Stat
+var unixMknod = unix.Mknod
+
+func addGadgetSlotDevice(chip *ChardevChip, instanceName, slotName string) error {
+	var stat unix.Stat_t
+	if err := unixStat(chip.Path, &stat); err != nil {
+		return err
+	}
+
+	devPath := SnapChardevPath(instanceName, slotName)
+	if err := os.MkdirAll(filepath.Dir(devPath), 0755); err != nil {
+		return err
+	}
+	// create a character device node for the slot, with major/minor numbers
+	// corresponding to the newly created aggregator device
+	if err := unixMknod(devPath, uint32(stat.Mode), int(stat.Rdev)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func removeGadgetSlotDevice(instanceName, slotName string) (aggregatedChip *ChardevChip, err error) {
+	devPath := SnapChardevPath(instanceName, slotName)
+	aggregatedChip, err = chardevChipInfo(devPath)
+	if err != nil {
+		return nil, err
+	}
+	return aggregatedChip, os.Remove(devPath)
+}
+
+func removeEphemeralUdevTaggingRule(gadget, slot string) error {
+	// XXX: is rule reload/trigger necessary
+	path := aggregatedChipUdevRulePath(gadget, slot)
+	return os.RemoveAll(path)
+}
+
+func removeAggregatedChip(aggregatedChip *ChardevChip) error {
+	// synchronize gpio helpers' access to the aggregator interface
+	unlocker, err := lockAggregator()
+	if err != nil {
+		return err
+	}
+	defer unlocker()
+
+	return os.WriteFile(filepath.Join(dirs.GlobalRootDir, aggregatorDeleteDevicePath), []byte(aggregatedChip.Label), 0644)
+}
+
+func validateLines(chip *ChardevChip, lines strutil.Range) error {
+	for _, span := range lines {
+		if uint(span.End) >= chip.NumLines {
+			return fmt.Errorf("invalid line offset %d: line does not exist in %q", span.End, chip.Name)
+		}
+	}
+
+	return nil
+}
+
+func findChips(filter func(chip *ChardevChip) bool) ([]*ChardevChip, error) {
+	allPaths, err := filepath.Glob(filepath.Join(dirs.DevDir, "/gpiochip*"))
```

Comment:
this may be too optimistic as it will match gpiochip devices we're created through aggregator. We should skip them